### PR TITLE
Feat/34 sort filter

### LINKED ISSUE #34: As a shopper, I can Sort and Filter the Product List
This ticket covers Sort & Filter functionality on the PLP and SRP. General UX as follows:

Image
Image

### PR DESCRIPTION
**Pull request by stephenrichter**
This PR adds basic sort functionality, including 2 new Flex UI widgets and a json_serializable extension.

Simulator Screenshot - iPhone 15 Pro Max - 2024-04-22 at 13 51 07

### CODE REVIEW

File: lib/flex_ui/widgets/pill.dart [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    super.key,
+    required this.text,
+    this.isSelected = false,
+    this.onTap,
+    this.margin = const EdgeInsets.only(left: 8),
+  });
+
+  final String text;
+  final bool isSelected;
+  final void Function()? onTap;
```
Comment (ncuillery) [inline lib/flex_ui/widgets/pill.dart]: Slightly more elegant: `final VoidCallback? onTap;`

File: lib/flex_ui/widgets/pill.dart
Comment (ncuillery) [inline lib/flex_ui/widgets/pill.dart:1]: Isn't it more or less the Material Chips component? https://m3.material.io/components/chips/overview

File: lib/product_list/cubits/product_list_cubit.dart [hunk 2]
```diff
@@ -30,4 +31,34 @@ class ProductListCubit extends Cubit<ProductListState> {
...
+    try {
+      emit(ProductListState(
+        status: Status.pending,
+        searchResults: state.searchResults,
+        products: state.products,
+      ));
+
+      final searchResults =
+          await GetIt.instance.get<ProductListApi>().fetchProducts(
+                categoryCode: state.searchResults!.currentQuery.lastLeaf,
```
Comment (ncuillery) [inline lib/product_list/cubits/product_list_cubit.dart]: Why not using the `categoryCode` in arguments?

File: lib/shared/utils/nested_reader.dart
Comment (ncuillery) [inline lib/shared/utils/nested_reader.dart:1]: Copy-pasting a large portion of code written by someone else inside the codebase has consequences. It became our duty to maintain it and fix bugs in it.

 It should be done in the last resort IMO, and always be associated to the question "is it worth it?"

In that case, all it does is sparing us to create a sub-model class. It brings more complex code to spare less dumb code.

Secondly: the only usage in the search results current query, an info we already have, we shouldn't map this field at all

File: lib/product_list/cubits/product_list_cubit.dart [hunk 3]
```diff
@@ -30,4 +31,34 @@ class ProductListCubit extends Cubit<ProductListState> {
...
+                sortBy: sortBy.code,
+              );
+      final products = searchResults.products;
```
Comment (ncuillery) [inline lib/product_list/cubits/product_list_cubit.dart]: `products` has a special treatment, why not destructuring the other fields in the cubit state?

File: lib/router.dart [hunk 4]
```diff
@@ -48,5 +52,14 @@ class AppRouter extends _$AppRouter {
...
           ],
         ),
+        ModalBottomSheetAutoRoute(
+          page: SortRoute.page,
+          path: '/sort',
+          shape: const RoundedRectangleBorder(
+            borderRadius: BorderRadius.vertical(
+              top: Radius.circular(FlexSizes.borderRadiusLg),
+            ),
+          ),
```
Comment (ncuillery) [inline lib/router.dart]: Strange to have UI declared here, is there a way to include that in the modal definition?

File: lib/flex_ui/layouts/modal_draggable_route.dart [hunk 5]
```diff
@@ -0,0 +1,102 @@
+import 'package:auto_route/auto_route.dart';
+import 'package:flex_storefront/flex_ui/layouts/modal_draggable_sheet.dart';
+import 'package:flutter/material.dart';
+
+/// A custom implementation of an [AutoRoute] that uses a [ModalBottomSheetRoute]
+/// to display a modal bottom sheet.
+///
+/// See https://github.com/Milad-Akarie/auto_route_library/issues/675
```
Comment (ncuillery) [inline lib/flex_ui/layouts/modal_draggable_route.dart]: The author of the gist said:

> If you don't want to use extra dependencies like the modal_bottom_sheet package, check this gist.

Do we not?

**Review by ncuillery — APPROVED** [review]

File: lib/flex_ui/widgets/pill.dart
Comment (stephenrichter) [reply lib/flex_ui/widgets/pill.dart:1]: Here's a good explanation/definitions with some research behind it haha
https://ux.stackexchange.com/questions/98069/what-is-the-difference-between-pills-chips-and-tags-tokens

File: lib/shared/utils/nested_reader.dart
Comment (stephenrichter) [reply lib/shared/utils/nested_reader.dart:1]: Let's discuss this one, I think it's worth it as I come across more randomly nested fields in Hybris - there is nothing to label these sub-model classes and accessing them vs. mapping them to the direct object is a pain.

**Pull request merged**